### PR TITLE
Fix MyPy errors in Apache Providers

### PR DESCRIPTION
**Pull request by potiuk**
Part of #19891

The .py additions are to handle "default_args" passed in
examples. Currently some of the obligatory parameters are
(correctly) passed as default_args. We have no good
mechanism yet to handle it properly for MyPy (it would
require to add a custom MyPy plugin to handle it)

We have no better way to handle it for now.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by potiuk** [comment]
cc: @khalidmammadov

**Comment by potiuk** [comment]
If we agree this is the right way of treating context we could easily automate and apply it to all operators/sensors that have currnetly:

```
execute(context)
```
or

```
poke(context)
```

as a single big change (after we complete all other MyPy changes. WDYT?

**Comment by potiuk** [comment]
```
[jarek:~/code/airflow] [airflow-master-3.7] fix-mypy-google-cloud-build+ ± grep -r "execute(self, context)" airflow/providers | wc
    362    1700   32196
[jarek:~/code/airflow] [airflow-master-3.7] fix-mypy-google-cloud-build+ ± grep -r "poke(self, context)" airflow/providers | wc
     21      86    1644
[jarek:~/code/airflow] [airflow-master-3.7] fix-mypy-google-cloud-build+ ± 
```

**Comment by potiuk** [comment]
And including airlfow built-ins (there we could directly import context):

```
[jarek:~/code/airflow] [airflow-master-3.7] fix-mypy-google-cloud-build+ 3s ± grep -r "execute(self, context)" airflow | wc
    373    1744   32907
[jarek:~/code/airflow] [airflow-master-3.7] fix-mypy-google-cloud-build+ ± grep -r "poke(self, context)" airflow | wc          
     28     114    2040
```

File: airflow/providers/apache/cassandra/example_dags/example_cassandra_dag.py [hunk 1]
```diff
@@ -35,7 +35,9 @@
     catchup=False,
     tags=['example'],
 ) as dag:
-    table_sensor = CassandraTableSensor(task_id="cassandra_table_sensor")
+    table_sensor = CassandraTableSensor(task_id="cassandra_table_sensor", table="table")
 
-    record_sensor = CassandraRecordSensor(task_id="cassandra_record_sensor", keys={"p1": "v1", "p2": "v2"})
+    record_sensor = CassandraRecordSensor(
+        task_id="cassandra_record_sensor", keys={"p1": "v1", "p2": "v2"}, table="table"
+    )
```
Comment (josh-fell) [inline airflow/providers/apache/cassandra/example_dags/example_cassandra_dag.py]: I think we should keep the `table` arg value in `default_args` and add the `type: ignore[call-arg]` comment instead. It seems like a proper fix is not going to be trivial but ignoring that specific Mypy error is a good interim solution.

The documentation for these sensors also references that the `table` parameter is set in `default_args`.

cc @uranusjr @khalidmammadov (related your Slack conversation)

Comment (potiuk) [reply airflow/providers/apache/cassandra/example_dags/example_cassandra_dag.py]: Maybe then we should make it simply optional and raise Exception if it is not set. That seems like the best approach and very simple to do.

Comment (potiuk) [reply airflow/providers/apache/cassandra/example_dags/example_cassandra_dag.py]: @josh-fell  - I pushed a fix with that. I thin it's simple and straightforward with Optional table and checking at runtime.

Comment (josh-fell) [reply airflow/providers/apache/cassandra/example_dags/example_cassandra_dag.py]: Random idea I just played around with that seemed for work here. What if we created a stub file for the example DAG?

`airflow/providers/apache/cassandra/example_dags/example_cassandra_dag.pyi`
```python
from typing import Any, Dict
from airflow.providers.apache.cassandra.hooks.cassandra import CassandraHook


class CassandraTableSensor:
    def __init__(
        self, *, table: str = ..., cassandra_conn_id: str = CassandraHook.default_conn_name, **kwargs: Any
    ) -> None:
        ...

class CassandraRecordSensor:
    def __init__(
        self,
        *,
        table: str = ...,
        keys: Dict[str, str] = ...,
        cassandra_conn_id: str = CassandraHook.default_conn_name,
        **kwargs: Any,
    ) -> None:
        ...


```
We don't have to change the operator logic or the example DAG itself and Mypy seems happy.

Comment (josh-fell) [reply airflow/providers/apache/cassandra/example_dags/example_cassandra_dag.py]: FYI - Just simplified the above stub file a little.

File: airflow/providers/apache/cassandra/sensors/record.py [hunk 2]
```diff
@@ -58,17 +63,19 @@ class CassandraRecordSensor(BaseSensorOperator):
...
-        table: str,
         keys: Dict[str, str],
+        table: str,
         cassandra_conn_id: str = CassandraHook.default_conn_name,
         **kwargs: Any,
     ) -> None:
         super().__init__(**kwargs)
         self.cassandra_conn_id = cassandra_conn_id
+        if table is None:
+            raise RuntimeError("Please make sure that you set `table` parameter in the sensor")
```
Comment (josh-fell) [inline airflow/providers/apache/cassandra/sensors/record.py]: If the stub files work, this can be removed from the sensors. Fingers crossed.

Comment (potiuk) [reply airflow/providers/apache/cassandra/sensors/record.py]: Ah yeah !

Comment (potiuk) [reply airflow/providers/apache/cassandra/sensors/record.py]: It works nicely with the stubs! Great idea!

**Comment by potiuk** [comment]
I am afraid I have no better idea for the "Context" imports from providers though :(

**Comment by potiuk** [comment]
Need some review here - anyone has better idea what to do instead of:

```
if TYPE_CHECKING:
    from airflow.utils.context import Context
else:
    Context = MutableMapping[str, Any]
```

I would like to automatically add it later to all the provider's executes/pokes to really benefit from the latest TypeDict Context improvements :D, but maybe we can find a better way ? The problem with it is we can't rely on the common code in airflow :(

**Comment by khalidmammadov** [comment]
Hi Jarek,

Since typing is not required for runtime, I have amended it to just below.

|if TYPE_CHECKING: from airflow.utils.context import Context|

Is that OK?

- Did the same in https://github.com/apache/airflow/pull/20409

**Comment by khalidmammadov** [comment]
Ignore me, we still need that part

**Comment by potiuk** [comment]
> Ignore me, we still need that part

Nice try though :)

**Comment by uranusjr** [comment]
I don’t think we need to `else` part. This should work:

```python
if TYPE_CHECKING:
    from airflow.utils.context import Context

def execute(self, context: "Context") -> None:
    ...
```

**Comment by josh-fell** [comment]
> I don’t think we need the `else` part. This should work:
> 
> ```python
> if TYPE_CHECKING:
>     from airflow.utils.context import Context
> 
> def execute(self, context: "Context") -> None:
>     ...
> ```

I can confirm this works in #20034 for the Docker decorator.

**Comment by potiuk** [comment]
Ah yeah! good one @uranusjr . I totally forgot that we can quote context :)

**Comment by potiuk** [comment]
That's why I was waiting with applying that. Now we can actually even make it truly "one-liner - cc: @khalidmammadov :

``if TYPE_CHECKING:  from airflow.utils.context import Context```

**Comment by potiuk** [comment]
Black is not happy with one-liner :(

**Comment by potiuk** [comment]
Should be good to go and I'd love t refresh stats after that :)

**Comment by uranusjr** [comment]
It seems I’m missing context on the pyi additions.

**Comment by khalidmammadov** [comment]
Looking at airflow/utils. Let me know if you as well

**Comment by potiuk** [comment]
> It seems I’m missing context on the pyi additions.

Context added.

**Comment by potiuk** [comment]
> It seems I’m missing context on the pyi additions.

Context added.

**Comment by potiuk** [comment]
Can I get the approval :) ? MyPy is calling.

**Review by eladkal — APPROVED** [review]

**Pull request merged**